torchseg/losses/dice.py

```diff
@@ -1,140 +1,147 @@
+            # Convert y_true to one_hot representation to compute DiceLoss
+            if num_classes == 1:
+                warnings.warn("Single channel prediction, 'mask_to_one_hot = True' ignored.")
+            else:
+                # maybe there is a better way to handle this?
+                permute_dims = tuple(dim - 1 for dim in spatial_dims)
+                y_true = F.one_hot(y_true, num_classes).squeeze(dim = 1) # N, 1, H, W, ... ---> N, H, W, ..., C
+                y_true = y_true.permute(0, -1, *permute_dims) # N, 1, H, W, ..., C ---> N, C, H, W, ...
+
+        if y_true.shape != y_pred.shape:
+            raise AssertionError(f"Ground truth has different shape ({y_true.shape})"
+                                 f" from predicted mask ({y_pred.shape})")
+
+        # Only reduce spatial dimensions
+        scores = soft_dice_score(y_pred,
+                                 y_true.type_as(y_pred),
+                                 power = self.power,
+                                 smooth = self.smooth,
+                                 eps = self.eps,
+                                 dims = spatial_dims)
 
         if self.log_loss:
             loss = -torch.log(scores.clamp_min(self.eps))
         else:
             loss = 1.0 - scores
 
-        # Dice loss is undefined for non-empty classes
+        # Dice loss is defined for non-empty classes
         # So we zero contribution of channel that does not have true pixels
         # NOTE: A better workaround would be to use loss term `mean(y_pred)`
         # for this case, however it will be a modified jaccard loss
-
-        mask = y_true.sum(dims) > 0
-        loss *= mask.to(loss.dtype)
+        """
+        Personal Opinion: maybe this can be skipped. By summing over batch
+        and spatial dimensions I think the idea is: I constantly have channels
+        with pixels = 0 which should not be included in the loss computation,
+        but if this happens there is something wrong with the masks.
+        While by just summing over spatial dimensions it means that an empty channel
+        should not be used to compute the DiceLoss, but it may simply mean
+        that the specific class is not present in this specific mask.
+        """
+        # to delete?
+        #dims = tuple(d for d in range(len(y_true.shape)) if d != 1)
+        #mask = y_true.sum(dims) > 0
+        #loss *= mask.to(loss.dtype)
 
         if self.classes is not None:
-            loss = loss[self.classes]
-
-        return self.aggregate_loss(loss)
+            loss = loss[:, self.classes, :]
 
-    def aggregate_loss(self, loss):
-        return loss.mean()
+        if self.reduction == LossReduction.MEAN:
+            loss = torch.mean(loss)
+        elif self.reduction == LossReduction.SUM:
+            loss = torch.sum(loss)
+        elif self.reduction == LossReduction.NONE:
+            broadcast_shape = list(loss.shape[0:2]) + [1] * (len(y_true.shape) - 2)
```

Comment:
What does this do?

Reply:
just giving the users the ability to choose a preferred reduction instead of the classic `mean`. The latter basically returns a DiceScore per channel. So, if y_true has shape [B, C, H, W], selecting `LossReduction.NONE` returns a loss with shape [B, C, 1, 1].